[LLM] Add llama 2 chatbot example (#2266)
skypilot-org/skypilot: llm/llama-v2/README.md

Reviewer: It seems like the stylization is "Llama 2". Should we update file name / content to reflect?

Author: Renamed. Thanks!

# Host Your Private LLaMA v2 Chatbot

The latest release of LLaMA v2 has been released with promising performance recently.

[LLaMA v2 release](https://github.com/facebookresearch/llama/tree/main)
[LLaMA v2 paper](https://ai.meta.com/research/publications/llama-2-open-foundation-and-fine-tuned-chat-models/)

## Pre-requisites

###  Step 1: Apply for the access to the LLaMA v2 model

Go to the [application page](https://ai.meta.com/resources/models-and-libraries/llama-downloads/) and apply for the access to the model weights.

### Step 2: Get the access token from huggingface

Generate a read-only access token on huggingface [here](https://huggingface.co/settings/token), and make sure your huggingface account can access the LLaMA v2 models [here](https://huggingface.co/meta-llama/Llama-2-7b-chat/tree/main).

Fill the access token in the [chatbot-hf.yaml](chatbot-hf.yaml) and [chatbot-meta.yaml](chatbot-meta.yaml) file.
```yaml
envs:
  MODEL_SIZE: 7
  HF_TOKEN: <your-huggingface-token>
```

## How to run LLaMA v2 chatbot (Huggingface model)?

You can now host your own LLaMA v2 chatbot with SkyPilot using 1-click.

> 1. Start the serving the LLaMA-7B-Chat v2 model on a single A100 GPU:

Reviewer 2: ```suggestion
1. Start serving the LLaMA-7B-Chat v2 model on a single A100 GPU:
```

Author: Done. Thanks!

```bash
sky launch -c llama-serve -s chatbot-hf.yaml
```

> 2. Check the output of the command. There will be a sharable gradio link (like the last line of the following). Open it in your browser to chat with Vicuna.

Reviewer 2: ```suggestion
2. Check the output of the command. There will be a sharable gradio link (like the last line of the following). Open it in your browser to chat with LLaMA v2.
```

Author: Done. Thanks1

# Host Your Private LLaMA v2 Chatbot

> The latest release of LLaMA v2 has been released with promising performance recently.

Reviewer 3: ```suggestion
[LLaMA 2](https://github.com/facebookresearch/llama/tree/main) is the top open-source models on the [Open LLM leaderboard](https://huggingface.co/spaces/HuggingFaceH4/open_llm_leaderboard) today. It has been released with a license that authorizes commercial use. You can deploy a private Llama 2 chatbot with SkyPilot in your own cloud with just one simple command.
```

Author: Done. Thanks!

* [LLaMA v2 release](https://github.com/facebookresearch/llama/tree/main)
* [LLaMA v2 paper](https://ai.meta.com/research/publications/llama-2-open-foundation-and-fine-tuned-chat-models/)

## Pre-requisites

###  Step 1: Apply for the access to the LLaMA v2 model

Go to the [application page](https://ai.meta.com/resources/models-and-libraries/llama-downloads/) and apply for the access to the model weights.

### Step 2: Get the access token from huggingface

Generate a read-only access token on huggingface [here](https://huggingface.co/settings/token), and make sure your huggingface account can access the LLaMA v2 models [here](https://huggingface.co/meta-llama/Llama-2-7b-chat/tree/main).

Fill the access token in the [chatbot-hf.yaml](chatbot-hf.yaml) and [chatbot-meta.yaml](chatbot-meta.yaml) file.
```yaml
envs:
  MODEL_SIZE: 7
  HF_TOKEN: <your-huggingface-token>
```

> ## How to run LLaMA v2 chatbot (Huggingface model)?

Reviewer 3: ```suggestion
## Running your own LLaMA 2 chatbot with SkyPilot
```

Author: Done. Thanks!

You can now host your own LLaMA v2 chatbot with SkyPilot using 1-click.

1. Start the serving the LLaMA-7B-Chat v2 model on a single A100 GPU:
```bash
sky launch -c llama-serve -s chatbot-hf.yaml
```

> 2. Check the output of the command. There will be a sharable gradio link (like the last line of the following). Open it in your browser to chat with Vicuna.

Reviewer 3: ```suggestion
2. Check the output of the command. There will be a sharable gradio link (like the last line of the following). Open it in your browser to chat with Llama 2.
```

Author: Done. Thanks!

```
(task, pid=20933) 2023-04-12 22:08:49 | INFO | gradio_web_server | Namespace(host='0.0.0.0', port=None, controller_url='http://localhost:21001', concurrency_count=10, model_list_mode='once', share=True, moderate=False)
(task, pid=20933) 2023-04-12 22:08:49 | INFO | stdout | Running on local URL:  http://0.0.0.0:7860
(task, pid=20933) 2023-04-12 22:08:51 | INFO | stdout | Running on public URL: https://<random-hash>.gradio.live
```

<p align="center">
  <img src="https://i.imgur.com/l5QX7D3.gif" alt="LLaMA v2 Demo"/>

> </p>

Reviewer 3: Gif doesn't seem to be loading in github preview for me..

Author: Yea... I am trying to make it work..

Author: Thanks to @romilbhardwaj, the GIF works now!

3. [Optional] Try other GPUs:
```bash
sky launch -c llama-serve-v100 -s chatbot-hf.yaml --gpus V100
```

4. [Optional] Serve the 13B model instead of the default 7B:
```bash
sky launch -c llama-serve -s chatbot-hf.yaml --env MODEL_SIZE=13

> ```

Reviewer 3: Have we tried the 70B model? If it requires a specific GPU (I guess A100-80GB), we should mention it here.

Author: Tested:
- [x] 7b on A100
- [x] 13b on A100
- [ ] 70b on A100-80GB

## How to run LLaMA v2 chatbot (FAIR model)?

> You can now host your own LLaMA v2 chatbot with SkyPilot using 1-click.

Reviewer 3: How is this different from the above? Would be nice to add a little more detail

Author: Changed the sentence. PTAL. : )
